crates/routes/src/nodeinfo.rs

```diff
@@ -84,6 +85,7 @@ struct NodeInfoWellKnownLinks {
   pub href: Url,
 }
 
+#[skip_serializing_none]
```

Comment:
They aren't allowed to be empty either.

All the fields are required in the spec.

```
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "id": "http://nodeinfo.diaspora.software/ns/schema/2.0#",
  "description": "NodeInfo schema version 2.0.",
  "type": "object",
  "additionalProperties": false,
  "required": [
    "version",
    "software",
    "protocols",
    "services",
    "openRegistrations",
    "usage",
    "metadata"
  ],
  ...
  ```
  see http://nodeinfo.diaspora.software/docson/index.html#/ns/schema/2.0#$$expand

Reply:
They are always set by Lemmy, only made optional to be able to receive data from other platforms which may not include them. Like the saying goes, "be conservative in what you send, be liberal in what you accept"

Reply:
Okay, but then I suggest that you return empty protocol list instead of leaving out the field.

Reply:
Looks like protocol, services, and metadata are the other ones that we don't have / send. The first two we could send empty lists, but metadata being required is a strange one.